### PR TITLE
[`feat`] Addition of popular image retrieval benchmark datasets

### PR DESCRIPTION
**Pull request by ir2718**
Hi,

this is a PR for issue https://github.com/KevinMusgrave/pytorch-metric-learning/issues/722. I've implemented four benchmark datasets: CUB-200, Cars196, INaturalist2018, and StanfordOnlineProducts. When using any of these datasets they will be downloaded directly and saved to the root directory, similar to the PyTorch dataset handling. Each of the implemented datasets inherits torch.utils.data.Dataset and can be used with dataloaders seamlessly. I've also added docs for each of the datasets implement, and a short overview of what users need to implemented if they want to add their own custom dataset. Tests for each of the datasets are also added. I've deliberately left out the `__init__.py` in `tests/datasets`, as each of the files has to be downloaded, and these can be pretty big (up to 130Gb).

@KevinMusgrave when you have time, please take a look and tell me if something requires changing.

### CODE REVIEW

**Comment by ir2718** [comment]
Are there any plans on merging this or would you like me to make some changes?

**Comment by KevinMusgrave** [comment]
Looks good I think, but it would be nice to get more "official" links for SOP and Cars, but especially for SOP because I don't like having people download an anonymous-looking google drive link 😅

**Comment by KevinMusgrave** [comment]
Thanks @ir2718!
